### PR TITLE
Add CI job to test compilation on Windows with vcpkg

### PR DESCRIPTION
**Pull request by traversaro**
Fix https://github.com/ros-industrial/abb_libegm/issues/92 .

The problems discussed https://github.com/ros-industrial/abb_libegm/issues/92 were actually just due to a wrong order of arguments passed to CMake (`cmake --config Release  --build .` seems to ignore the `--build` flag). 

The correct compilation was tested in https://github.com/traversaro/abb_libegm/pull/2 .
The CI uses https://github.com/marketplace/actions/run-vcpkg to cache the installation of vcpkg dependencies, ensuring that the build on the master branch (after the first one, that is used to populate the cache) do not take ~20 minutes.

### CODE REVIEW

File: .github/workflows/windows-vcpkg.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    - uses: actions/checkout@master
+    
+    # Restore from cache the previously built ports. If "cache miss"
+    # then provision vcpkg, install desired ports, finally cache everything for the next run.
+    - name: Restore from cache and run vcpkg
+      uses: lukka/run-vcpkg@v3
+      with:
+        vcpkgArguments: '--triplet x64-windows boost-asio boost-math boost-smart-ptr protobuf'
+        # commit from vcpkg's master branch on 2020/06/01
+        vcpkgGitCommitId: 6d36e2a86baf8d227fc6dce587bd69997d67fb5e
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:26]: Does this take care of the "use vcpkg as a submodule of your repository" recommendation without using a sub module?

File: .github/workflows/windows-vcpkg.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        vcpkgDirectory: '${{ github.workspace }}/../vcpkg'
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:27]: The example [here](https://github.com/marketplace/actions/run-vcpkg#setup-vcpkg-and-install-ports) seems to use a different path. Does our repository require something special here?

File: .github/workflows/windows-vcpkg.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      with:
+        vcpkgArguments: '--triplet x64-windows boost-asio boost-math boost-smart-ptr protobuf'
+        # commit from vcpkg's master branch on 2020/06/01
+        vcpkgGitCommitId: 6d36e2a86baf8d227fc6dce587bd69997d67fb5e
+        vcpkgDirectory: '${{ github.workspace }}/../vcpkg'  
+
+    - name: Configure the project 
+      shell: bash
+      run: | 
+        mkdir  build
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:32]:
```suggestion
        mkdir build
```

File: .github/workflows/windows-vcpkg.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        cd build
+        cmake -DCMAKE_TOOLCHAIN_FILE=${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake ..
+        cmake --build . --config Release
+        cmake --install . --config Release
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:36]: I know we're not debugging anything, but could we perhaps make this `RelWithDebInfo` (unless `Release` is much better/required)?

In my experience ppl will start copy-pasting these commands to their own build files/command windows and having debug symbols available would make stack traces much easier to understand.

File: .github/workflows/windows-vcpkg.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
+name: Windows vcpkg Workflow
+
+on:
+  push:
+  pull_request:
+  schedule:
+  # * is a special character in YAML so you have to quote this string
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:7]: Remove.

File: .github/workflows/windows-vcpkg.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+name: Windows vcpkg Workflow
+
+on:
+  push:
```
Comment (gavanderhoorn) [inline .github/workflows/windows-vcpkg.yml:4]:
```suggestion
  push:
    branches:
      - master
```
?

Just so we avoid building everything twice when we push a branch directly to the repo to create a PR.

Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml:4]: It make sense.

File: .github/workflows/windows-vcpkg.yml (hunk 1 above)
Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml:28]: No, it just  downloads and compiles (or extract from the cache, if available) vcpkg in the directory specified in `vcpkgDirectory`. 

In general I do not think that "use vcpkg as a submodule of your repository" is a recommendation valid  for a general purpose project. I think that tracking in an extremely rigorous and reproducible way dependencies (also using version control, as when using git submodules) make sense for projects that do not need to be reused, such as closed source software that is meant for a specific deployment or use case. But for open source projects that are meant to be reused in multiple projects as this library, mixing  the project itself and how to get dependencies I do not think it is helpful.

File: .github/workflows/windows-vcpkg.yml (hunk 2 above)
Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml:29]: Yes, it is related to this line https://github.com/ros-industrial/abb_libegm/blob/dcff5647b77a7da5fb80d689dcabd3366b135894/CMakeLists.txt#L91 . In particular, if you try to configure that line with recent CMake while storing the installation prefix of the dependencies inside the source directory, you will get this error: 
~~~
CMake Error in CMakeLists.txt:
  Target "abb_libegm" INTERFACE_INCLUDE_DIRECTORIES property contains path:

    "D:/a/abb_libegm/abb_libegm/vcpkg/installed/x64-windows/include"

  which is prefixed in the source directory.
~~~
basically it is an early fatal error against non-relocatable config files.
We can avoid this workaround  once we will link protobuf via imported CMake targets.

Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml:30]: In https://github.com/ros-industrial/abb_libegm/pull/101/commits/fe944c4e5f1f5eb27de063d13772ccf3ab1734cb I added a brief explanation.

File: .github/workflows/windows-vcpkg.yml (hunk 5 above)
Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml]: Done in https://github.com/ros-industrial/abb_libegm/pull/101/commits/fe944c4e5f1f5eb27de063d13772ccf3ab1734cb .

File: .github/workflows/windows-vcpkg.yml (hunk 4 above)
Comment (traversaro) [reply .github/workflows/windows-vcpkg.yml]: Done in https://github.com/ros-industrial/abb_libegm/pull/101/commits/fe944c4e5f1f5eb27de063d13772ccf3ab1734cb .

**Comment by traversaro** [comment]
@gavanderhoorn I should have addressed all the comments and cleanup the commit history.

**Comment by gavanderhoorn** [comment]
I'm assuming we can duplicate this for `abb_librws`?

Or would the Poco dependency be difficult to set up?

**Comment by traversaro** [comment]
> 
> 
> I'm assuming we can duplicate this for `abb_librws`?
> 
> Or would the Poco dependency be difficult to set up?

Fortunately `poco` is available as a vcpkg port (see https://github.com/microsoft/vcpkg/tree/master/ports/poco), so it should be possible to just port this to `abb_librws` and changing the installed dependencies.

**Comment by jontje** [comment]
Nice work! 😄 🎉
